### PR TITLE
std.fmt.format — with streams

### PR DESCRIPTION
**Pull request by fengb**
Spiritual successor to https://github.com/ziglang/zig/pull/4059. Thanks to @daurnimator for suggesting this approach.

Like generators, we can coalesce context/output/error into one argument.

Unlike generators, we don't have to invent a new pattern. OutStream is already an existing standard, and we just have to leverage its functionality. This also circumvents the previously encountered generator footguns since we have at most 1 async context.

~The major downside is that OutStream is currently "all or nothing" when it comes to async, which would prevent non-async version of format(). Not sure how much of a problem this is and it's not specific to just `format()`, and this could be solvable by introducing noasync OutStreams (yay colors?)~
*solved via https://github.com/ziglang/zig/pull/4710*

### CODE REVIEW

File: lib/std/buffer.zig [hunk 1]
```diff
@@ -149,7 +143,27 @@ pub const Buffer = struct {
     }
 
     pub fn print(self: *Buffer, comptime fmt: []const u8, args: var) !void {
-        return std.fmt.format(self, error{OutOfMemory}, Buffer.append, fmt, args);
+        const BufferStream = struct {
```
Comment (fengb) [inline lib/std/buffer.zig]: This is indicating we should have a standard way of converting Buffer into an OutStream.

Comment (komuw) [reply lib/std/buffer.zig]: https://github.com/ziglang/zig/pull/4665/files#diff-d55cbcbbfea0b63c681837c0d91763d5R212-R231

**Comment by MasterQ32** [comment]
> The major downside is that OutStream is currently "all or nothing" when it comes to async, which would prevent non-async version of format(). Not sure how much of a problem this is and it's not specific to just format(), and this could be solvable by introducing noasync OutStreams (yay colors?)

I think this can be a major problem when doing user interfaces. I use `std.fmt` primarily for user interaction and UI formatting. Requiring `std.fmt.format` to be async because it's using streams would mean that even just setting the text of a label would require to be async. Sure one can use `noasync`, but i actually like the current approach to format.

**Comment by fengb** [comment]
@MasterQ32 I think the new OutStream rewrite solves the always-async problem, and it exposes that the concept of an OutStream is just `Context, Error, writeFn` glued together into one.

File: lib/std/buffer.zig [hunk 2]
```diff
@@ -155,7 +151,7 @@ pub const Buffer = struct {
     }
 
     pub fn print(self: *Buffer, comptime fmt: []const u8, args: var) !void {
-        return std.fmt.format(self, error{OutOfMemory}, Buffer.append, fmt, args);
+        return self.outStream().print(fmt, args);
```
Comment (fengb) [inline lib/std/buffer.zig]: Now that the stream is readily available, maybe this method should be deleted.

**Comment by andrewrk** [comment]
This idea is approved :heavy_check_mark:  but I haven't reviewed the PR yet

**Pull request merged**